### PR TITLE
prevent loading other extensions when precompiling an extension

### LINKED ISSUE #55557: Extension cycles are problematic (long `precompile` times, difficult to understand/resolve)
The attached test package imports many packages, including ModelingToolkit which currently has a circular dependency among its extensions. In a case with circular deps like this, `pkg> precompile` seems to be much less successful on 1.10.5 (which back-ported circularity detection) vs. 1.10.4

A locally-built v1.10.5 shows the warning regarding the circularity:
```julia
(Circularity_MWE) pkg> precompile
┌ Warning: Circular dependency detected. Precompilation will be skipped for:
│   ModelingToolkitStandardLibrary [16a59e39-deab-5bd0-87e4-056b12336739]
│   NonlinearSolve [8913a72c-1f9b-4ce2-8d82-65094dcecaec]
...
│   BoundaryValueDiffEq [764a87c0-6b3e-53db-9096-fe964310641d]
│   ManifoldsBoundaryValueDiffEqExt [eb713886-0b96-5c41-a09d-5e8967e02f7d]
└ @ Pkg.API ~/repos/julia/usr/share/julia/stdlib/v1.10/Pkg/src/API.jl:1279
...
  342 dependencies successfully precompiled in 294 seconds. 31 skipped due to circular dependency.
```

The `using Circularity_MWE` message is quite scary and cryptic:
```julia
┌ Warning: Module SymbolicsPreallocationToolsExt with build ID ffffffff-ffff-ffff-0000-6f7fcf83bfa3 is missing from the cache.
│ This may mean SymbolicsPreallocationToolsExt [d479e226-fb54-5ebe-a75e-a7af7f39127f] does not support precompilation but is imported by a module that does.
└ @ Base loading.jl:1948
┌ Error: Error during loading of extension SymbolicsPreallocationToolsExt of Symbolics, use `Base.retry_load_extensions()` to retry.
│   exception =
│    1-element ExceptionStack:
│    Declaring __precompile__(false) is not allowed in files that are being precompiled.
│    Stacktrace:
│      [1] _require(pkg::Base.PkgId, env::Nothing)
│        @ Base ./loading.jl:1952
... (many stack frames skipped)
```

The real problem though, is it takes about **25 minutes** for the first `using X` to complete on my machine:
```julia
$ time julia --project=Circularity_MWE -e "using Circularity_MWE"
 2076.34s user 31.37s system 139% cpu 24:53.46 total
```

On v1.10.4, this takes ~10 seconds.

**MWE:** [Circularity_MWE.tar.gz](https://github.com/user-attachments/files/16698188/Circularity_MWE.tar.gz)

Related: #55543

### PR DESCRIPTION
**Pull request by KristofferC**
The current way of loading extensions when precompiling an extension very easily leads to cycles. For example, if you have more than one extension and you happen to transitively depend on the triggers of one of your extensions you will immediately hit a cycle where the extensions will try to load each other indefinitely. This is an issue because you cannot directly influence your transitive dependency graph so from this p.o.v the current system of loading extension is "unsound".

The test added here checks this scenario and we can now precompile and load it without any warnings or issues.

Would have made https://github.com/JuliaLang/julia/issues/55517 a non issue.

Fixes https://github.com/JuliaLang/julia/issues/55557

### CODE REVIEW

**Comment by KristofferC** [comment]
@nanosoldier `runtests()`

**Comment by nanosoldier** [comment]
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/55589#issuecomment-2310399221) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/80f0764_vs_6477530/report.html) is available.

**Comment by KristofferC** [comment]
@dlfivefifty, you seem to be the only one using this feature (at least in PkgEval) in e.g. https://github.com/JuliaArrays/LazyArrays.jl/blob/7777232394f9e42a7685cbf15001bf4b3bcba696/ext/LazyArraysBlockBandedMatricesExt.jl#L25.

To check, would it be possible and how hard would it be to provide the same functionality without requiring this. The reason is that extensions loading extensions (as it is done now) is quite brittle and easily lead to cycles and bad experience. There might be a slightly more "relaxed" version of this that could be implemented that could maybe still allow the use case in LazyArrays but I want to check with you if the simple thing could be done first.

**Comment by dlfivefifty** [comment]
I felt this was a bit dodgy when I did it. But I don’t know how else to do this?

note this usage should never lead to cycles since BlockBandedMatrices.jl itself depends on BandedMatrices.jl

**Comment by dlfivefifty** [comment]
Put another way: if package B depends on A, then it’s natural that an extension for B depends on an extension on A

**Comment by KristofferC** [comment]
> note this usage should never lead to cycles since BlockBandedMatrices.jl itself depends on BandedMatrices.jl

If you happen to get two of your triggers for your extensions (say BlockBandedMatrices and StaticArrays) into your full dependency graph of LazyArrays you get a cycle because loading LazyArrays will trigger one extension and loading LazyArrays in that will trigger the other (over and over).

You cannot control your full dependency graph so this can happen at any time.

**Comment by dlfivefifty** [comment]
I don't understand your example. What's the alternative solution?

It might be possible I can move functions/types into LazyArrays.jl in this case. But it seems like a major flaw in extensions combining with packages.

**Comment by KristofferC** [comment]
> But it seems like a major flaw in extensions combining with packages

Well yeah, that's what we are trying to fix heh.

**Comment by dlfivefifty** [comment]
I still don’t understand the issue being solved. A directed graph doesn’t form cycles….

**Comment by KristofferC** [comment]
https://github.com/JuliaLang/julia/issues/52511#issuecomment-1983388150 is one example in the wild.

You can also just setup the test example I provided in this PR and see how precompilation fails for example.

**Comment by dlfivefifty** [comment]
I think I need a mathematical description to be convinced it’s not an implementational issue.

Packages with (or without) extensions are simply a tree. How does a cycle come up? 

I how are extensions different to before when we could make packages to implement extra behaviour? Ie how is LazyArrays making extensions for BandedMatrices and BlockBandedMatrixes in any way different to making two packages called LazyArraysBandedMatrices and LazyArraysBlockBandedMatrices that depend on each other exactly like the extensions do??

**Comment by vtjnash** [comment]
Mainly because the edges point the unexpected direction from BandedMatrices and LazyArrays, since loading either of those can trigger loading the extension, so in the graph, they are the originators of the edge, rather than the terminators. That can cause a cycle to appear in the load graph that isn't as easily anticipated.

**Comment by topolarity** [comment]
I've created https://github.com/JuliaArrays/LazyArrays.jl/pull/345 to avoid the inter-ext dependency

This kind of workaround isn't always possible though - This wouldn't work if the types I moved (e.g. `AbstractLazyBandedLayout`) had fields/supertypes from both BandedMatrices.jl (the trigger) and LazyArrays.jl

I think we need to make the change that Kristoffer described here: https://github.com/JuliaLang/julia/issues/48734#issuecomment-1554626135 and give this situation a definite load-/dependency-ordering, so that it's not a "cycle" any more - With that change, I think the only "cycle" left is for extensions that share the exact same trigger/parent set (e.g. `extBA` and `extAB`) for which case this fix probably makes sense

**Comment by topolarity** [comment]
OK, I think we probably want to add an exception to this to still allow loading of other extensions if their triggers are a strict subset of your own (e.g. `{A,C} ⊊ {A,B,C}`)

That'd leave room for us to do something like https://github.com/JuliaLang/julia/pull/49891 to allow `ext -> ext` dependencies, like the LazyArrays.jl case.

**Comment by dlfivefifty** [comment]
Could a simple solution be that if an extensionA depends on extensionB it needs to list it in the Project.toml? I.e. 

https://github.com/JuliaArrays/LazyArrays.jl/blob/d3e51078551d98ba951525931c77b2f259bfea29/Project.toml#L25

Would become
```
LazyArraysBlockBandedMatricesExt = ["LazyArraysBandedMatricesExt", "BlockBandedMatrices"]
```

File: base/precompilation.jl [hunk 1]
```diff
@@ -471,7 +471,10 @@ function precompilepkgs(pkgs::Vector{String}=String[];
                     extension_1 == extension_2 && continue
                     deps_ext_2 = depsmap_transitive[extension_2]
                     if issubset(deps_ext_2, deps_ext_1)
-                        push!(depsmap[extension_1], extension_2)
+                        # extensions do not depend on each other
+                        if !haskey(exts, extension_2)
```
Comment (topolarity) [inline base/precompilation.jl:475]: When is `extension_2` not in `exts`?

**Review by topolarity — APPROVED** [review]
Thanks @KristofferC !

This doesn't give us a solution for direct `ext → ext` dependencies, but we can add that piece in later.

For now, this is much better behavior 👍

**Pull request merged**